### ISSUE TITLE
SourceLocation as a relative path 

### ISSUE THREAD

**Issue by kadel** [openshift/odo#1552]
```
▶ odo create java --binary target/spring-boot-rest-http-1.5.17-5-SNAPSHOT.jar
 ✓  Checking component
 ✓  Checking component version
Please use `odo push` command to create the component with source deployed

▶ cat .odo/config.yaml
kind: LocalConfig
apiversion: odo.openshift.io/v1alpha1
ComponentSettings:
  Type: java
  SourceLocation: /Users/tkral/tmp/odo-examples/spring-boot-http-booster/target/spring-boot-rest-http-1.5.17-5-SNAPSHOT.jar
  SourceType: binary
  Application: spring-boot-hekh
  Project: default
  Name: java-olhp

```

```
▶ odo create java
 ✓  Checking component
 ✓  Checking component version
Please use `odo push` command to create the component with source deployed

▶ cat .odo/config.yaml
kind: LocalConfig
apiversion: odo.openshift.io/v1alpha1
ComponentSettings:
  Type: java
  SourceLocation: /Users/tkral/tmp/odo-examples/spring-boot-http-booster
  SourceType: local
  Application: spring-boot-fqkv
  Project: default
  Name: java-yaiq
```

SourceLocation in a config needs to be saved as a relative path to the context directory.

In the above example `SourceLocation`  should be `target/spring-boot-rest-http-1.5.17-5-SNAPSHOT.jar` for the binary component and  `./` or `.`  for the local component.

We can't put the whole path there because it prenets sharing the config

**Comment by kadel** [comment]
@girishramnani this needs to addressed before beta release

**Comment by girishramnani** [comment]
Do you want me to take this?

**Comment by kadel** [comment]
> Do you want me to take this?

that would be great

**Comment by kadel** [comment]
@girishramnani this might be related to something that @cdrage  is doing in his PR https://github.com/openshift/odo/pull/1446

**Comment by girishramnani** [comment]
Yeah, seems so

**Comment by kadel** [comment]
any progress here?

**Comment by cdrage** [comment]
I commented here: https://github.com/openshift/odo/issues/1429 as both of these issues are relevant.

I am currently investigating this (got distracted by doc changes I've been pinged on), at the moment, it looks like I'll have to add another `FromSlash` somewhere in the code.

**Comment by kadel** [comment]
> For issue #1429: I am currently investigating this (got distracted by doc changes I've been pinged on), at the moment, it looks like I'll have to add another `FromSlash` somewhere in the code.

I think  that it should be `ToSlash` not `From` ;-)

**Comment by cdrage** [comment]
Been distracted by issue #1429 as I'm trying to fix the Windows ignore bug.

Question, with regards to storing the actual filepath, should we be storing it as per whatever OS is being used, or ALWAYS forward slash?

**Comment by kadel** [comment]
> Been distracted by issue #1429 as I'm trying to fix the Windows ignore bug.

When you fix this issue, the #1429 will get resolved.

> Question, with regards to storing the actual filepath, should we be storing it as per whatever OS is being used, or ALWAYS forward slash?

always forward slash!

It needs to be stored in forward slash format. Before the path is stored, it needs to be properly converted to relative path, and piped through `ToSlash`. This will ensure that it is always stored in a unified way (with forwarding slashes). When it is read it will be piped through `FromSlash` to ensure that it is converted to current OS format
